src/Microsoft.ML.Tokenizers/Model/Model.cs

```diff
@@ -16,83 +16,147 @@ public abstract class Model
+        {
+            if (maxTokens <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxTokens), "The maximum number of tokens must be greater than 0.");
+            }
+
+            var ids = new List<int>();
+
+            if (maxTokens == int.MaxValue)
             {
-                accumulatedIds.Add(token.Id);
+                EncodeToIds(text, ids, out _);
+                textLength = text.Length;
+                return ids.Count;
             }
+
+            IReadOnlyList<Token> tokens = Encode(text);
+            textLength = 0;
+            int count = Math.Min(tokens.Count, maxTokens);
+            for (int i = 0; i < count; i++)
+            {
+                textLength += tokens[i].Offset.Length;
+            }
+
+            return count;
         }
 
         /// <summary>
         /// Get the number of tokens that the input text will be encoded to.
         /// </summary>
-        /// <param name="text">The text to encode. If the value of the parameter <paramref name="isSpecialToken"/> is true, the entire text will be treated as a special token.</param>
-        /// <param name="isSpecialToken">Specifies whether the entire <paramref name="text"/> is considered a special token.</param>
+        /// <param name="text">The text to encode.</param>
+        /// <param name="textIndex">Starting from this index to the end of the text will encompasses the maximum encoded tokens.</param>
+        /// <param name="maxTokens">The maximum number of tokens to encode.</param>
         /// <returns>The number of tokens that the input text will be encoded to.</returns>
         /// <remarks>
         /// This method does the default implementation that uses the EncodeToIds method to get the number of token's Ids.
         /// Tokenizer's models which care about performance may choose to override this method to provide a more efficient implementation.
         /// </remarks>
-        public virtual int CountTokens(ReadOnlySpan<char> text, bool isSpecialToken)
+        public virtual int CountTokensFromEnd(ReadOnlySpan<char> text, out int textIndex, int maxTokens = int.MaxValue)
         {
+            if (maxTokens <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxTokens), "The maximum number of tokens must be greater than 0.");
+            }
+
             var ids = new List<int>();
-            EncodeToIds(text, isSpecialToken, ids);
-            return ids.Count;
+
+            if (maxTokens == int.MaxValue)
+            {
+                EncodeToIds(text, ids, out _);
+                textIndex = 0;
+                return ids.Count;
+            }
+
+            IReadOnlyList<Token> tokens = Encode(text);
+            textIndex = text.Length;
+            int count = Math.Min(tokens.Count, maxTokens);
+            for (int i = tokens.Count - 1; i >= tokens.Count - count; i--)
```

Comment:
Nit: this condition is going to make an interface call on every iteration. Better to store it into a local first and then use that local.
```C#
int tokensCount = tokens.Count;
int end = tokensCount - count;
for (int i = tokensCount - 1; i >= end; i--)
```